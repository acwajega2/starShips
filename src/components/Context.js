import React, { Component } from 'react';
import axios from 'axios' //--for perforning http requests

import Data from './Data'





//----Creating a starShip Context
const StarShipContext = React.createContext();

/*
1.We shall be using the StarShipContext context API to pass data across all the components using th StarShip Provider
2. We shall wrap the StarShip provider on top of our application

*/ 


//----> Creating the StartShipProvider

class StarShipProvider extends Component{

//----Creating/declaring the state
constructor (props){
    super(props)
    this.state ={
        starShips : [],
        err_msg : "",
        //----pagination variables
        pageNumbers: [],
        currentPage: 1,
        itemsPerPage: 3,
        totalItems : 0,
        totalPages : 0,
        hasLogedIn : false,
        currentUser: {},
        nextPage_Url: "",
        prevPage_Url: ""
        
    }
}
//----End of state declaration

componentDidMount(){
    /* */

    //----> Getting thr starShips on Component Mounting
    this.getStarShips()
    //this.initialiseStarShips()

    //IntialiseStarShips
}


componentDidUpdate(){
    /* */
}


componentWillUnmount(){
/* */
}


//  >>>>>>>>>>>>>>>>>>>   Implementing Functions to be used
initialiseStarShips = () =>{
    this.setState(() =>{
        return {
            starShips : Data.response,
            totalItems: Data.response.length,
            totalPages : Math.floor(Data.response.length / this.state.itemsPerPage)
            
        }
    })
    
  }


   
    
    

//-------> Get StarShips function
getStarShips = () =>{
    //----> we shall use axios to perform a get request and it will the return a promise
    axios.get("https://swapi.co/api/starships/")
    .then(res => {
        console.log(res.data.results)
        //----> Setting the starShip property of the starte
        this.setState(() =>{
            return {
                starShips: res.data.results,
                nextPage_Url: res.data.next,
                prevPage_Url:res.data.previous,
                totalItems: res.data.results.length,
                itemsPerPage: res.data.results.length,
                totalPages : Math.floor(res.data.results.length / this.state.itemsPerPage)
            }
        })

    })
}

    
//---------Function to navigate to the next webpage
    
    nextPage_link = () => {

        if (this.state.nextPage_Url !== null) {
            axios.get(this.state.nextPage_Url)
            .then(res => {
                console.log(res.data.results)
                //----> Setting the starShip property of the starte
                this.setState(() =>{
                    return {
                        starShips: res.data.results,
                        nextPage_Url: res.data.next,
                        prevPage_Url: res.data.previous,
                        itemsPerPage: res.data.results.length,
                        totalItems: res.data.results.length,
                        totalPages : Math.floor(res.data.results.length / this.state.itemsPerPage)
                    }
                })
        
            })

        }
    
}

    
//-------------------End of Next webpage

    
//-----------------Navigating to the previous page
    prevPage_link = () => {
        if (this.state.prevPage_Url !== null) {
            axios.get(this.state.prevPage_Url)
            .then(res => {
                console.log(res.data.results)
                //----> Setting the starShip property of the starte
                this.setState(() =>{
                    return {
                        starShips: res.data.results,
                        nextPage_Url: res.data.next,
                        prevPage_Url:res.data.previous,
                        totalItems: res.data.results.length,
                        itemsPerPage: res.data.results.length,
                        totalPages : Math.floor(res.data.results.length / this.state.itemsPerPage)
                    }
                })
        
            })

        }
        
    }
//--------------------End of Previous Page
showMessage =() =>{
    console.log("someone clicked changePage")

}

//------> Handle Social Login Success
handleSocialLogin =(user) => {

    console.log(user.profile)

    this.setState(()=>{
        return{
            hasLogedIn: true,
            user : user.profile,
            currentUser : user.profile

        }

    })

}


//--------



//---------> Handle Social Login Failure
handleSocialLoginFailure =(err) =>{
    this.setState(()=>{
        return{
            hasLogedIn: false,
            user : []

        }

    })

}



//--------


//-------->  Changing a page-----
changePage =(page) =>{       
    this.setState( () =>{
        return {
            currentPage: parseInt(page),
            err_msg : "sdsdsd",
        }
    }
   
    )
  }
//------- End of Changing page

//------> NextPage 
nextPage = () =>{
       this.state.currentPage !== this.state.totalPages + 1?
       this.setState( () =>{      
        return {
           currentPage : Number(this.state.currentPage) + 1,          
            err_msg : "sdsdsd",
        }
    })
        : 
        this.setState( () =>{      
            return {
               currentPage : Number(this.state.currentPage) ,              
                err_msg : "sdsdsd",
            }
        })
 }



//------> Previous Page
prevPage = () =>{
    this.state.currentPage !== 1?
    this.setState( () =>{   
     return {
        currentPage : Number(this.state.currentPage) - 1,       
         err_msg : "sdsdsd",
     }
 })
     : 
     this.setState( () =>{   
         return {
            currentPage : Number(this.state.currentPage) ,           
             err_msg : "sdsdsd",
         }
     })

}


//-----> Go to Home Page -----
homePage = () =>{
    this.setState( () =>{   
        return {
           currentPage : 1 ,           
            err_msg : "sdsdsd",
        }
    })

}
// >>>>>>>>> End of Functions

render(){

    return(
        <StarShipContext.Provider value={{...this.state, 
        getStarShips : this.getStarShips,
        changePage :this.changePage,
        showMessage :this.showMessage,
        nextPage : this.nextPage,
        prevPage : this.prevPage,
        homePage : this.homePage,    
        handleSocialLogin : this.handleSocialLogin,
            handleSocialLoginFailure: this.handleSocialLoginFailure,
            prevPage_link: this.prevPage_link,
            nextPage_link : this.nextPage_link
        
        }}>
             {this.props.children}
        </StarShipContext.Provider>
    )
}




}



//----->  Creating the Startship Consumer
const StarShipConsumer = StarShipContext.Consumer;


//------> So we shall export both the starship Provider and Consumer
export {StarShipProvider,StarShipConsumer}