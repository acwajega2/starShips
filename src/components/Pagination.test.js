import React from 'react';
import { shallow, mount } from '../Enzyme';
import XPagination from './Pagination';
import renderer from 'react-test-renderer';
import {StarShipProvider} from './Context'


describe("Testing the Pagination Component", () =>{


 //------Doing SnapShot Testing---------------------------
    it("Matches the previous snap shot", () => {
     
    const component = renderer.create( < StarShipProvider ><XPagination  />      
        </StarShipProvider>).toJSON();
    expect(component).toMatchSnapshot()
    
});
//------- End of Snapshot Testing--------------------



  //-----App Initializing Test-----
  it("Should Run Properly",()=>{
    shallow(<XPagination />)
  });
  //-----End of Initializing Test----
    
    
    

    
  
 

})