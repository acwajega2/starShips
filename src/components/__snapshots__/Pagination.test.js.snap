exports[`Testing the Pagination Component Matches the previous snap shot 1`] = `
<ul
  className="ant-pagination"
  style={Object {}}
  unselectable="unselectable">
  <li
    aria-disabled={true}
    className="ant-pagination-disabled ant-pagination-prev"
    onClick={[Function]}
    onKeyPress={[Function]}
    tabIndex={null}
    title="Previous Page">
    <a
      className="ant-pagination-item-link">
      <i
        aria-label="icon: left"
        className="anticon anticon-left"
        onClick={undefined}
        tabIndex={undefined}>
        <svg
          aria-hidden="true"
          className=""
          data-icon="left"
          fill="currentColor"
          focusable="false"
          height="1em"
          onClick={undefined}
          style={undefined}
          viewBox="64 64 896 896"
          width="1em">
          <path
            d="M724 218.3V141c0-6.7-7.7-10.4-12.9-6.3L260.3 486.8a31.86 31.86 0 0 0 0 50.3l450.8 352.1c5.3 4.1 12.9.4 12.9-6.3v-77.3c0-4.9-2.3-9.6-6.1-12.6l-360-281 360-281.1c3.8-3 6.1-7.7 6.1-12.6z" />
        </svg>
      </i>
    </a>
  </li>
  <li
    className="ant-pagination-item ant-pagination-item-0 ant-pagination-disabled ant-pagination-item-disabled"
    onClick={[Function]}
    onKeyPress={[Function]}
    tabIndex="0"
    title={0}>
    <a>
      0
    </a>
  </li>
  <li
    aria-disabled={true}
    className="ant-pagination-disabled ant-pagination-next"
    onClick={[Function]}
    onKeyPress={[Function]}
    tabIndex={null}
    title="Next Page">
    <a
      className="ant-pagination-item-link">
      <i
        aria-label="icon: right"
        className="anticon anticon-right"
        onClick={undefined}
        tabIndex={undefined}>
        <svg
          aria-hidden="true"
          className=""
          data-icon="right"
          fill="currentColor"
          focusable="false"
          height="1em"
          onClick={undefined}
          style={undefined}
          viewBox="64 64 896 896"
          width="1em">
          <path
            d="M765.7 486.8L314.9 134.7A7.97 7.97 0 0 0 302 141v77.3c0 4.9 2.3 9.6 6.1 12.6l360 281.1-360 281.1c-3.9 3-6.1 7.7-6.1 12.6V883c0 6.7 7.7 10.4 12.9 6.3l450.8-352.1a31.96 31.96 0 0 0 0-50.4z" />
        </svg>
      </i>
    </a>
  </li>
</ul>
`;
